src/bidiMapper/domains/context/PreloadScriptStorage.ts

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2023 Google LLC.
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import crypto from 'crypto';
+
+import {CommonDataTypes, Script} from '../../../protocol/protocol';
+
+import {CdpTarget} from './cdpTarget';
+
+export type BidiPreloadScript = {
+  /** Browsing context ID. */
+  contextId: CommonDataTypes.BrowsingContext | null;
+  /** BiDi ID, an automatically generated UUID. */
+  id: string;
+  /** CDP preload scripts. */
+  cdpPreloadScripts: CdpPreloadScript[];
+  /** The script itself, in a format expected by the spec i.e. a function. */
+  scriptSource: string;
+};
+
+/** BidiPreloadScripts can be filtered by either context ID or BiDi ID. */
+export type BidiPreloadScriptFilter = Partial<
+  Pick<BidiPreloadScript, 'contextId'> & Pick<BidiPreloadScript, 'id'>
+>;
+
+export type CdpPreloadScript = {
+  /** CDP target. */
+  target: CdpTarget;
+  /** CDP preload script ID. */
+  preloadScriptId: Script.PreloadScript;
+};
+
+/**
+ * Container class for preload scripts.
+ *
+ * BiDi IDs are generated by the server and are unique within the context.
+ *
+ * CDP preload script IDs are generated by the client and are unique
+ * within the session.
+ *
+ * The mapping between BiDi and CDP preload script IDs is 1:many.
+ * BiDi IDs are needed by the mapper to keep track of potential multiple CDP IDs
+ * in the client.
+ *
+ * This class does not concern itself with the validity of the IDs.
+ */
+export class PreloadScriptStorage {
```

Comment:
There should be a method to add a CDP preload script to the `BidiPreloadScript`.

Reply:
To an existing one, right?

Reply:
correct